keps/0020-controller-overhaul.md

```diff
@@ -0,0 +1,160 @@
+More complex manager pod deployment with additional service
+How do we manage RBAC?
+
+### Have a separate CRD just for maintaining commands to execute a plan
+
+This would basically mean going half-way back to state pre KEP-18 where we had a separate CRD that was created for every plan that was executed. In this implementation we could have a much leaner CRD capturing only the requests to execute a plan with very lean status that does not duplicate the instance's status. Could be as simple as:
+
+```
+apiVersion: kudo.dev/v1beta1
+kind: PlanExecutionRequest
+Spec:
+  planName: deploy
+Status:
+  Started: true
+```
+
+The question is whether we would add a Status into that CRD. If not, it would be hard to tell the client if the execution was even possible and how it ended (in comparison to the API solution where we can validate as part of HTTP response and even return some unique ID for tracking). If we decide to include status, we’re back in pre-KEP-18 world where keeping those two Statuses in sync is hard in kubernetes world with no transactions and both CRDs being reconciled in their own loops.
+
+But at the same time this seems to be the way people are leaning toward when coming to similar use-cases. The discussion under [this issue](https://github.com/kubernetes/kubernetes/issues/72637#issuecomment-515566586) is kind of touching on the topic of Request objects. What this [KEP](https://github.com/metal3-io/metal3-docs/pull/48/files) proposes for reboot is also kind of similar to what we’re trying to do. Alongside what they’re proposing we might not even need a controller for this CRD and we could treat it as “queue” meaning every request will be fulfilled so e.g. when you create a request and a plan is running, instance will pick up the next request when it has a capacity to run another plan.
+
+If we do that, it would mean:
+- there is no controller for PlanExecutionRequest
+- you cannot tell status of the execution just from the PlanExecution object
+- InstanceController will manage lifecycle of PlanExecutionRequest and will be responsible for garbage collection
+
+#### Pros
+
+Kubernetes-native way of API for executing plans, can be easily done via kubectl or any other kubernetes client
+Seems to currently be the de-facto pattern to do these things before sub-resources via webhooks are introduced for CRDs.
+
+#### Cons
+
+Challenging to communicate back to the user if execution was possible and how it happened (would need a good support for this in CLI)
+Have to maintain another CRD
+Need to think about garbage collection of the Request objects
+
+### Introduce new sub-resource for triggering plan executions
+
+The ideal solution for this would be if custom sub-resources for CRDs are supported. That is not the case right now, see discussion under [this issue](https://github.com/kubernetes/kubernetes/issues/72637).
```

Comment:
It won't help us today but could we maybe push this issue? /cc @gerred

Reply:
good point, I asked about it in sig-api-machinery slack :)

Reply:
Subresource would be great if they were all stored in different etcd keys. I could see a plan execution history (especially once we have event-based triggers and parameters as part of plan executions) easily getting to 1MB plus, and we'd have to write code to limit that, while with CRDs we can always just sweep for older ones and delete the whole CRD.